pkg/webhook/v1alpha2/applicationconfiguration/validating_handler.go

```diff
@@ -67,97 +80,117 @@ func (h *ValidatingHandler) Handle(ctx context.Context, req admission.Request) a
+	for _, c := range v.validatingComps {
+		if c.appConfigComponent.ComponentName != "" && c.appConfigComponent.RevisionName != "" {
+			allErrs = append(allErrs, fmt.Errorf(errFmtRevisionName,
+				c.appConfigComponent.ComponentName, c.appConfigComponent.RevisionName))
 		}
 	}
-	return true, ""
+	return allErrs
 }
 
-// checkWorkloadNameForVersioning check whether versioning-enabled component workload name is empty
-func checkWorkloadNameForVersioning(ctx context.Context, client client.Reader, dm discoverymapper.DiscoveryMapper,
-	appConfig *v1alpha2.ApplicationConfiguration) (bool, string) {
-	for _, v := range appConfig.Spec.Components {
-		acc := v
-		vEnabled, err := checkComponentVersionEnabled(ctx, client, dm, &acc)
-		if err != nil {
-			return false, fmt.Sprintf(errFmtCheckWorkloadName, err.Error())
-		}
-		if !vEnabled {
-			continue
+// ValidateWorkloadNameForVersioningFn validates workload name for version-enabled component
+func ValidateWorkloadNameForVersioningFn(_ context.Context, _ client.Client, _ discoverymapper.DiscoveryMapper,
+	v ValidatingAppConfig) []error {
+	var allErrs []error
+	for _, c := range v.validatingComps {
+		for _, t := range c.validatingTraits {
+			if t.traitDefinition.Spec.RevisionEnabled {
+				goto Validate
+			}
 		}
-		c, _, err := util.GetComponent(ctx, client, acc, appConfig.GetNamespace())
-		if err != nil {
-			return false, fmt.Sprintf(errFmtCheckWorkloadName, err.Error())
+		continue
+	Validate:
+		if ok, workloadName := checkParams(c.component.Spec.Parameters, c.appConfigComponent.ParameterValues); !ok {
+			allErrs = append(allErrs, fmt.Errorf(errFmtWorkloadNameNotEmpty, workloadName))
 		}
-
-		if ok, workloadName := checkParams(c.Spec.Parameters, acc.ParameterValues); !ok {
-			return false, fmt.Sprintf(reasonFmtWorkloadNameNotEmpty, workloadName)
+		if workloadName := c.workloadContent.GetName(); workloadName != "" {
+			allErrs = append(allErrs, fmt.Errorf(errFmtWorkloadNameNotEmpty, workloadName))
 		}
+	}
+	return allErrs
+}
 
-		w := &fieldpath.Paved{}
-		if err := json.Unmarshal(c.Spec.Workload.Raw, w); err != nil {
-			return false, fmt.Sprintf(errFmtUnmarshalWorkload, c.GetName(), err.Error())
-		}
-		workload := unstructured.Unstructured{Object: w.UnstructuredContent()}
-		workloadName := workload.GetName()
-
-		if len(workloadName) != 0 {
-			return false, fmt.Sprintf(reasonFmtWorkloadNameNotEmpty, workloadName)
+// ValidateTraitAppliableToWorkloadFn validates whether a trait is allowed to apply to the workload.
+func ValidateTraitAppliableToWorkloadFn(_ context.Context, _ client.Client, _ discoverymapper.DiscoveryMapper, v ValidatingAppConfig) []error {
+	klog.Info("validate applicationConfiguration", "name", v.appConfig.Name)
+	var allErrs []error
+	for _, c := range v.validatingComps {
+		workloadType := c.component.GetLabels()[oam.WorkloadTypeLabel]
+		workloadDefRefName := c.workloadDefinition.Spec.Reference.Name
+		// TODO version of workloadDefRef should be taken into account
```

Comment:
what does it mean?

Reply:
Since [definition is allowed to have more than one version](https://github.com/crossplane/oam-kubernetes-runtime/blob/2cf8f6353b3c40eecdbe915bf199ec0544d848b9/apis/core/v1alpha2/core_types.go#L34), maybe ApplyTo is possible to indicate a specific version of WorkloadDefinition in the future.